File: src/main/java/uk/co/threebugs/mochiwhattotrade3/trade/TradeService.java
package uk.co.threebugs.mochiwhattotrade3.trade;

import lombok.AllArgsConstructor;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.springframework.data.domain.Example;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import uk.co.threebugs.mochiwhattotrade3.SlackClient;
import uk.co.threebugs.mochiwhattotrade3.Type;
import uk.co.threebugs.mochiwhattotrade3.account.Account;
import uk.co.threebugs.mochiwhattotrade3.accountsetupgroups.AccountSetupGroups;
import uk.co.threebugs.mochiwhattotrade3.accountsetupgroups.AccountSetupGroupsRepository;
import uk.co.threebugs.mochiwhattotrade3.metatrader.Client;
import uk.co.threebugs.mochiwhattotrade3.metatrader.Order;
import uk.co.threebugs.mochiwhattotrade3.metatrader.TradeInfo;
import uk.co.threebugs.mochiwhattotrade3.setup.SetupRepository;
import uk.co.threebugs.mochiwhattotrade3.utils.Constants;
import uk.co.threebugs.mochiwhattotrade3.utils.TimeHelper;

import java.math.BigDecimal;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.util.List;
import java.util.Optional;

import static java.time.ZoneOffset.UTC;
import static uk.co.threebugs.mochiwhattotrade3.setup.SetupFileRepository.getNextEventTime;

@AllArgsConstructor
@Service
public class TradeService {

    private static final Logger logger = LogManager.getLogger(TradeService.class);

    private final TradeRepository tradeRepository;
    private final SetupRepository setupRepository;
    private final AccountSetupGroupsRepository accountSetupGroupsRepository;
    private final TradeMapper tradeMapper;
    private final TimeHelper timeHelper;
    private final SlackClient slackClient;

    public static BigDecimal addTicks(BigDecimal initialPrice, int ticksToAdd, BigDecimal tickSize) {
        var ticks = tickSize.multiply(BigDecimal.valueOf(ticksToAdd));
        return initialPrice.add(ticks);
    }

    public Optional<TradeDto> findById(int id) {
        return tradeRepository.findById(id)
                              .map(tradeMapper::toDto);
    }

    public TradeDto save(TradeDto tradeDto) {

        var setup = this.setupRepository.findById(tradeDto.getSetup()
                                                          .getId())
                                        .orElseThrow(() -> new RuntimeException("Setup not found"));

        var record = tradeMapper.toEntity(tradeDto, setup);
        record = tradeRepository.save(record);
        return tradeMapper.toDto(record);
    }

    public void deleteById(int id) {
        tradeRepository.deleteById(id);
    }

    public List<TradeDto> findAll() {


        return this.tradeRepository.findAll()
                                   .stream()
                                   .map(tradeMapper::toDto)
                                   .toList();
    }

    public List<TradeDto> findTrades(TradeDto exampleRecord, Sort sort) {

        var example = Example.of(tradeMapper.toEntity(exampleRecord));

        return tradeRepository.findAll(example, sort)
                              .stream()
                              .map(tradeMapper::toDto)
                              .toList();
    }

    public void createTradesToPlaceFromEnabledSetups(String symbol, Account account) {

        var accountSetupGroups = accountSetupGroupsRepository.findByAccount(account);

        accountSetupGroups.forEach(accountSetupGroup -> createAndPlaceTradesForEnabledSetupsInSetupGroup(symbol, account, accountSetupGroup));

    }

    private void createAndPlaceTradesForEnabledSetupsInSetupGroup(String symbol, Account account, AccountSetupGroups accountSetupGroup) {
        var setups = setupRepository.findEnabledSetups(symbol, accountSetupGroup.getSetupGroups());
        setups.forEach(setup -> {
            var placedDateTime = getNextEventTime(setup.getDayOfWeek(), setup.getHourOfDay());
            tradeRepository.findBySetupAndPlacedDateTimeAndAccount(setup, placedDateTime, account)
                           .ifPresentOrElse(trade -> {
                               // logger.info("Trade already exists: " + trade);
                           }, () -> {
                               var trade = tradeMapper.toEntity(setup, placedDateTime, account);
                               trade.setType(Type.WAITING_TO_PLACED);
                               tradeRepository.save(trade);
                               slackClient.sendSlackNotification(trade.getNewTradeMessage());
                           });

        });
    }

    public void placeTrades(Client dwx, String symbol, BigDecimal bid, BigDecimal ask, Account account) {


        tradeRepository.findByTypeAndSetup_SymbolAndAccount(Type.WAITING_TO_PLACED, symbol, account)
                       .forEach(trade -> {
                           var now = ZonedDateTime.now(UTC);
                           var placedDateTime = trade.getPlacedDateTime();
                           if (now.isAfter(placedDateTime)) {
                               tradeRepository.save(placeTrade(dwx, bid, ask, trade));
                           }
                       });


    }

    public Trade placeTrade(Client dwx, BigDecimal bid, BigDecimal ask, Trade trade) {
        var fillPrice = trade.getSetup()
                             .isLong() ? ask : bid;

        var orderType = trade.getSetup()
                             .isLong() ? "buylimit" : "selllimit";

        var tickSize = new BigDecimal("0.00001");
        if (trade.getSetup()
                 .getSymbol()
                 .equalsIgnoreCase(Constants.USDJPY)) {
            tickSize = new BigDecimal("0.01");
        }

        var price = addTicks(fillPrice, trade.getSetup()
                                             .getTickOffset(), tickSize);
        var stopLoss = addTicks(fillPrice, trade.getSetup()
                                                .getStop(), tickSize);
        var takeProfit = addTicks(fillPrice, trade.getSetup()
                                                  .getLimit(), tickSize);


        //        dwx.openOrder(Order.builder()
//                           .symbol("EURUSD")
//                           .orderType("buylimit")
//                           .lots(0.01)
//                           .price(new BigDecimal("1.02831"))
//                           .stopLoss(new BigDecimal("1.00831"))
//                           .takeProfit(new BigDecimal("1.0931"))
//                           .magic(100)
//                           .comment("test")
//                           .expiration(TradeService.addSecondsToCurrentTime(8))
//                           .build());


        var magic = trade.getId();
        dwx.openOrder(Order.builder()
                           .symbol(trade.getSetup()
                                        .getSymbol())
                           .orderType(orderType)
                           .lots(0.01)
                           .price(price)
                           .stopLoss(stopLoss)
                           .takeProfit(takeProfit)
                           .magic(magic)
                           .comment(trade.getSetup()
                                         .concatenateFields())
                           .expiration(timeHelper.addSecondsToCurrentTime(trade.getSetup()
                                                                               .getOutOfTime()))
                           .build());

        trade.setType(Type.PLACED);

        slackClient.sendSlackNotification("Order placed: " + trade.getSetup()
                                                                  .concatenateFields());
        return trade;
    }

    public void closeTradesAtTime(Client dwx, String symbol, Account account) {


        tradeRepository.findByTypeAndSetup_SymbolAndAccount(Type.FILLED, symbol, account)
                       .stream()
                       .filter(record -> record.getPlacedDateTime()
                                               .plusHours(record.getSetup()
                                                                .getTradeDuration())
                                               .isBefore(ZonedDateTime.now(UTC)))
                       .forEach(trade -> {
                           dwx.closeOrdersByMagic(trade.getId());
                           trade.setType(Type.CLOSED_BY_TIME);
                           trade.setClosedDateTime(ZonedDateTime.now());
                           tradeRepository.save(trade);

                           slackClient.sendSlackNotification("Order closed: " + trade.getSetup()
                                                                                     .getRank() + " " + trade.getSetup()
                                                                                                             .getSymbol() + " " + (trade.getSetup()
                                                                                                                                        .isLong() ? "LONG" : "SHORT") + " " + trade.getProfit());

                       });


    }

    public void fillTrade(TradeInfo tradeInfo, Integer metatraderId) {

        tradeRepository.findById(tradeInfo.getMagic())
                       .ifPresentOrElse(trade -> {
                                   trade.setPlacedPrice(tradeInfo.getOpenPrice());
                                   trade.setPlacedDateTime(ZonedDateTime.of(tradeInfo.getOpenTime(), ZoneId.of("Europe/Zurich")));
                                   trade.setType(Type.PLACED_IN_MT);
                                   trade.setMetatraderId(metatraderId);

                                   tradeRepository.save(trade);

                           slackClient.sendSlackNotification("Order placed in MT: " + trade);
                               }, () -> logger.info("Trade not found: " + tradeInfo)

                       );
    }

    public void closeTrade(TradeInfo tradeInfo) {
        tradeRepository.findById(tradeInfo.getMagic())
                       .ifPresentOrElse(trade -> closeTrade(tradeInfo, trade), () -> logger.info("Trade not found: " + tradeInfo));

    }

    private void closeTrade(TradeInfo tradeInfo, Trade trade) {
        trade.setType(Type.CLOSED_BY_USER);
        trade.setClosedPrice(tradeInfo.getTakeProfit());
        trade.setClosedDateTime(ZonedDateTime.now());
        trade.setProfit(tradeInfo.getProfitAndLoss());

        tradeRepository.save(trade);

        slackClient.sendSlackNotification("Order closed: %d %s %s %s".formatted(trade.getSetup()
                                                                                     .getRank(), trade.getSetup()
                                                                                                      .getSymbol(), trade.getSetup()
                                                                                                                         .isLong() ? "LONG" : "SHORT", trade.getProfit()));
    }
}

File: src/main/kotlin/uk/co/threebugs/darwinexclient/trade/TradeService.kt
package uk.co.threebugs.darwinexclient.trade

import org.springframework.data.domain.Example
import org.springframework.data.domain.Sort
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import uk.co.threebugs.darwinexclient.SlackClient
import uk.co.threebugs.darwinexclient.Status
import uk.co.threebugs.darwinexclient.accountsetupgroups.AccountSetupGroups
import uk.co.threebugs.darwinexclient.clock.MutableClock
import uk.co.threebugs.darwinexclient.metatrader.Client
import uk.co.threebugs.darwinexclient.metatrader.Order
import uk.co.threebugs.darwinexclient.metatrader.TradeInfo
import uk.co.threebugs.darwinexclient.setup.Setup
import uk.co.threebugs.darwinexclient.setup.SetupFileRepository
import uk.co.threebugs.darwinexclient.setup.SetupRepository
import uk.co.threebugs.darwinexclient.utils.Constants
import uk.co.threebugs.darwinexclient.utils.TimeHelper
import uk.co.threebugs.darwinexclient.utils.logger
import java.math.BigDecimal
import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.util.function.Consumer

@Service
class TradeService(
    private val tradeRepository: TradeRepository,
    private val tradeMapper: TradeMapper,
    private val timeHelper: TimeHelper,
    private val slackClient: SlackClient,
    private val clock: MutableClock,
    private val setupRepository: SetupRepository,
) {
    fun findById(id: Int): TradeDto? {
        return tradeRepository.findByIdOrNull(id)?.let { tradeMapper.toDto(it) }
    }


    fun save(tradeDto: TradeDto): TradeDto {
        val setup = findSetupById(tradeDto.setup!!.id!!)

        return tradeMapper.toEntity(tradeDto, setup, clock)
            .let(tradeRepository::save)
            .let(tradeMapper::toDto)
    }

    private fun findSetupById(id: Int): Setup {
        return setupRepository.findByIdOrNull(id)
            ?: throw IllegalArgumentException("Setup not found for ID: $id")
    }

    fun deleteById(id: Int) {
        tradeRepository.deleteById(id)
    }

    fun deleteTradesBySetupGroupsName(name: String): Int {
        return tradeRepository.deleteBySetupGroupsName(name)
    }

    fun findAll(): List<TradeDto> {
        return tradeRepository.findAll().filterNotNull().map { tradeMapper.toDto(it) }
    }

    fun findTrades(exampleRecord: TradeDto, sort: Sort): List<TradeDto?> {
        val example = Example.of(tradeMapper.toEntity(exampleRecord, clock))

        return tradeRepository.findAll(example, sort).map { tradeMapper.toDto(it) }
    }

    fun createTradesToPlaceFromEnabledSetups(symbol: String, accountSetupGroups: AccountSetupGroups) {

        val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")


        val setups = setupRepository.findEnabledSetups(symbol, accountSetupGroups.setupGroups!!)
        setups.forEach { setup ->
            val targetPlaceTime: ZonedDateTime =
                SetupFileRepository.getNextEventTime(setup.dayOfWeek!!, setup.hourOfDay!!, clock)
            //logger.info("clockNow: ${ZonedDateTime.now()} targetPlaceTime: ${formatter.format(targetPlaceTime)}" + " setup: ${setup.id}" + " accountSetupGroups: ${accountSetupGroups.id}")
            val existingTrade =
                tradeRepository.findBySetupAndTargetPlaceDateTimeAndAccountSetupGroups(
                    accountSetupGroups.id!!,
                    setup.id!!,
                    formatter.format(targetPlaceTime)
                )
            if (existingTrade == null) {
                val trade = tradeMapper.toEntity(setup, targetPlaceTime, accountSetupGroups.account!!, clock).apply {
                    status = Status.PENDING
                }
                tradeRepository.save(trade)
                slackClient.sendSlackNotification(trade.newTradeMessage)
            }
        }
    }


    fun placeTrades(
        dwx: Client,
        symbol: String,
        bid: BigDecimal,
        ask: BigDecimal,
        accountSetupGroups: AccountSetupGroups
    ) {
        tradeRepository.findByAccountSetupGroupsSymbolAndStatus(accountSetupGroups.id!!, symbol, Status.PENDING.name)
            .forEach(Consumer { trade: Trade ->
                val now = ZonedDateTime.now(clock)
                val targetPlaceDateTime = trade.targetPlaceDateTime
                if (now.isAfter(targetPlaceDateTime)) {
                    tradeRepository.save(placeTrade(dwx, bid, ask, trade))
                }
            })
    }

    fun placeTrade(dwx: Client, bid: BigDecimal, ask: BigDecimal, trade: Trade): Trade {
        val fillPrice = if (trade.setup!!.isLong) ask else bid
        val orderType = if (trade.setup!!.isLong) "buylimit" else "selllimit"
        var tickSize = BigDecimal("0.00001")
        if (trade.setup!!.symbol.equals(Constants.USDJPY, ignoreCase = true)) {
            tickSize = BigDecimal("0.01")
        }
        val price = addTicks(fillPrice, trade.setup!!.tickOffset!!, tickSize)
        val stopLoss = addTicks(fillPrice, trade.setup!!.stop!!, tickSize)
        val takeProfit = addTicks(fillPrice, trade.setup!!.limit!!, tickSize)


        //        dwx.openOrder(Order.builder()
//                           .symbol("EURUSD")
//                           .orderType("buylimit")
//                           .lots(0.01)
//                           .price(new BigDecimal("1.02831"))
//                           .stopLoss(new BigDecimal("1.00831"))
//                           .takeProfit(new BigDecimal("1.0931"))
//                           .magic(100)
//                           .comment("test")
//                           .expiration(TradeService.addSecondsToCurrentTime(8))
//                           .build());
        val magic = trade.id
        dwx.openOrder(
            Order(
                symbol = trade.setup!!.symbol,
                orderType = orderType,
                lots = 0.01,
                price = price,
                stopLoss = stopLoss,
                takeProfit = takeProfit,
                magic = magic,
                expiration = timeHelper.addSecondsToCurrentTime(trade.setup!!.outOfTime!!.toLong()),
                comment = trade.setup!!.concatenateFields()
            )
        )

        trade.status = Status.ORDER_SENT
        slackClient.sendSlackNotification("Order placed: " + trade.setup!!.concatenateFields())
        return trade
    }

    fun closeTradesAtTime(dwx: Client, symbol: String, accountSetupGroups: AccountSetupGroups) {
        tradeRepository.findByAccountSetupGroupsSymbolAndStatus(accountSetupGroups.id!!, symbol, Status.FILLED.name)
            .stream()
            .filter { trade: Trade ->

                val closeDateTime = trade.targetPlaceDateTime!!.plusHours(trade.setup!!.tradeDuration!!.toLong())
                val currentDateTime = ZonedDateTime.now(clock)

                //logger.info("closeDateTime: $closeDateTime currentDateTime: $currentDateTime")

                closeDateTime.isBefore(currentDateTime)
            }.forEach { trade: Trade ->
                dwx.closeOrdersByMagic(trade.id!!)
                trade.status = Status.CLOSED_BY_MAGIC_SENT
                //trade.closedDateTime = ZonedDateTime.now()
                tradeRepository.save(trade)
                slackClient.sendSlackNotification("Order closed by magic: ${trade.setup!!.rank} ${trade.setup!!.symbol} ${trade.setup!!.direction} ${trade.profit}")
            }
    }

    fun fillTrade(tradeInfo: TradeInfo, metatraderId: Int) {
        tradeRepository.findByIdOrNull(tradeInfo.magic)?.let { trade ->
            if (trade.status == Status.PENDING || trade.status == Status.ORDER_SENT) {
                trade.apply {
                    placedPrice = tradeInfo.openPrice
                    placedDateTime = ZonedDateTime.of(tradeInfo.openTime, ZoneId.of("Europe/Zurich"))
                    status = Status.PLACED_IN_MT
                    this.metatraderId = metatraderId
                }.also {
                    tradeRepository.save(it)
                    slackClient.sendSlackNotification("Order placed in MT: $it")
                }
            }
        } ?: logger.warn("Trade not found: $tradeInfo")
    }

    fun closeTrade(tradeInfo: TradeInfo) {
        tradeRepository.findByIdOrNull(tradeInfo.magic)?.let { trade ->
            closeTrade(
                tradeInfo,
                trade
            )
        } ?: logger.warn("Trade not found: $tradeInfo")
    }

    private fun closeTrade(tradeInfo: TradeInfo, trade: Trade) {

        val closingStatus =

            if (trade.status == Status.CLOSED_BY_MAGIC_SENT) {
                Status.CLOSED_BY_TIME
            } else if (tradeInfo.type.equals("buy") || tradeInfo.type.equals("sell")) {
                Status.CLOSED_BY_USER
            } else if (tradeInfo.type.equals("buylimit") || tradeInfo.type.equals("selllimit")) {
                Status.OUT_OF_TIME
            } else {
                Status.CLOSED_BY_USER
            }

        trade.apply {
            status = closingStatus
            closedPrice = tradeInfo.takeProfit
            closedDateTime = ZonedDateTime.now(clock)
            profit = tradeInfo.profitAndLoss
        }.also { tradeRepository.save(it) }

        slackClient.sendSlackNotification("Order closed: ${trade.setup!!.rank} ${trade.setup!!.symbol} ${trade.setup!!.direction} ${trade.profit}")
    }

    fun deleteTradesByAccountName(name: String): Int {
        return tradeRepository.deleteByAccountName(name)
    }

    fun findByAccountName(name: String): List<TradeDto> {
        return tradeRepository.findByAccount_Name(name).map { tradeMapper.toDto(it) }
    }

    fun findBySetupGroupsName(name: String): List<TradeDto>? {
        return tradeRepository.findBySetupGroupsName(name).map { tradeMapper.toDto(it) }
    }

    companion object {
        fun addTicks(initialPrice: BigDecimal, ticksToAdd: Int, tickSize: BigDecimal): BigDecimal {
            val ticks = tickSize.multiply(BigDecimal.valueOf(ticksToAdd.toLong()))
            return initialPrice.add(ticks)
        }
    }
}
